parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecFactory.java

```diff
@@ -0,0 +1,364 @@
+      return BytesInput.from(output);
+    }
+
+    @Override
+    public void decompress(ByteBuffer input, int compressedSize, ByteBuffer output, int uncompressedSize)
+        throws IOException {
+
+      decompressor.reset();
+      byte[] inputBytes = new byte[compressedSize];
+      input.position(0);
+      input.get(inputBytes);
+      decompressor.setInput(inputBytes, 0, inputBytes.length);
+      byte[] outputBytes = new byte[uncompressedSize];
+      decompressor.decompress(outputBytes, 0, uncompressedSize);
+      output.clear();
+      output.put(outputBytes);
+    }
+
+    @Override
+    protected void release() {
+      DirectCodecPool.INSTANCE.returnDecompressor(decompressor);
+    }
+  }
+
+  public class FullDirectDecompressor extends DirectBytesDecompressor {
+    private final Object decompressor;
+    private ByteBuffer compressedBuffer;
+    private ByteBuffer uncompressedBuffer;
+    private ExposedHeapBytesDecompressor extraDecompressor;
+    public FullDirectDecompressor(CompressionCodec codec){
+      this.decompressor = DirectCodecPool.INSTANCE.codec(codec).borrowDirectDecompressor();
+      this.extraDecompressor = new ExposedHeapBytesDecompressor(codec);
+    }
+
+    @Override
+    public BytesInput decompress(BytesInput compressedBytes, int uncompressedSize) throws IOException {
+    	return extraDecompressor.decompress(compressedBytes, uncompressedSize);
+    }
+
+    @Override
+    public void decompress(ByteBuffer input, int compressedSize, ByteBuffer output, int uncompressedSize)
+        throws IOException {
+      output.clear();
+      try {
+        DirectCodecPool.DECOMPRESS_METHOD.invoke(decompressor, (ByteBuffer) input.limit(compressedSize), (ByteBuffer) output.limit(uncompressedSize));
+      } catch (IllegalAccessException e) {
+        e.printStackTrace();
+      } catch (InvocationTargetException e) {
+        e.printStackTrace();
+      }
+      output.position(uncompressedSize);
+    }
+
+    @Override
+    protected void release() {
+      compressedBuffer = DirectCodecFactory.this.release(compressedBuffer);
+      uncompressedBuffer = DirectCodecFactory.this.release(uncompressedBuffer);
+      DirectCodecPool.INSTANCE.returnDecompressor(decompressor);
+      extraDecompressor.release();
+    }
+
+  }
+
+  public class NoopDecompressor extends DirectBytesDecompressor {
+
+    @Override
+    public void decompress(ByteBuffer input, int compressedSize, ByteBuffer output, int uncompressedSize)
+        throws IOException {
+      Preconditions.checkArgument(compressedSize == uncompressedSize,
+          "Non-compressed data did not have matching compressed and uncompressed sizes.");
+      output.clear();
+      output.put((ByteBuffer) input.duplicate().position(0).limit(compressedSize));
+    }
+
+    @Override
+    public BytesInput decompress(BytesInput bytes, int uncompressedSize) throws IOException {
+      return bytes;
+    }
+
+    @Override
+    protected void release() {
+    }
```

Comment:
This now extends just the single abstract base class, so this method does need to be implemented